database/src/ops/block.rs

```diff
@@ -1,89 +1,398 @@
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV1,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV1s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn block_exists() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v1<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV1)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV1s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
 }
 
 /// TODO
-pub fn get_block_hash() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block_v2<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV2,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV2s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn get_block_height() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v2<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV2)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV2s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
 }
 
 /// TODO
-pub fn get_block_weight() {
-    todo!()
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_block_v3<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height: BlockHeight,
+    block: &BlockInfoV3,
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    env.open_db_rw::<BlockInfoV3s>(tx_rw)?.put(&height, block)
 }
 
 /// TODO
-pub fn get_block_already_generated_coins() {
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn add_blocks_v3<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+    height_and_blocks: &[(BlockHeight, BlockInfoV3)],
+) -> Result<(), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    let mut table = env.open_db_rw::<BlockInfoV3s>(tx_rw)?;
+    for (height, block) in height_and_blocks {
+        table.put(height, block)?;
+    }
+    Ok(())
+}
+
+/// TODO
+///
+/// # Errors
+/// TODO
+#[inline]
+pub fn pop_block<'env, Ro, Rw, Env>(
+    env: &Env,
+    tx_rw: &mut Rw,
+) -> Result<(BlockHeight, BlockInfoLatest), RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    todo!(); // remove related info?
+    env.open_db_rw::<BlockInfoV3s>(tx_rw)?.pop_last()
+}
```

Comment:
Not exactly sure what else to remove in our tables.

Reply:
This should remove everything in the block, all block info, all txs in the blocks and all tx info from the DB